File: Adapters/AdMob/Classes/SAAdMobAdapter.m
#import "SAAdMobAdapter.h"
#import "SAAdMobExtras.h"
#import "SAAdMobRewardedAd.h"
#import "SAAdMobBannerAd.h"
#import "SAAdMobInterstitialAd.h"
#include <stdatomic.h>

#define kERROR_DOMAIN @"tv.superawesome.SAAdMobAdapter"

@implementation SAAdMobAdapter {
    SAAdMobRewardedAd *_rewardedAd;
    SAAdMobBannerAd *_bannerAd;
    SAAdMobInterstitialAd *_interstitialAd;
}

+ (GADVersionNumber) adSDKVersion {
    NSString *versionString = [[AwesomeAds info] versionNumber];
    NSArray *versionComponents = [versionString componentsSeparatedByString:@"."];
    GADVersionNumber version = {0};
    if (versionComponents.count == 3) {
        version.majorVersion = [versionComponents[0] integerValue];
        version.minorVersion = [versionComponents[1] integerValue];
        version.patchVersion = [versionComponents[2] integerValue];
    }
    return version;
}

+ (GADVersionNumber) adapterVersion {
    NSString *versionString = [[AwesomeAds info] versionNumber];
    NSArray *versionComponents = [versionString componentsSeparatedByString:@"."];
    GADVersionNumber version = {0};
    if (versionComponents.count == 3) {
        version.majorVersion = [versionComponents[0] integerValue];
        version.minorVersion = [versionComponents[1] integerValue];
        version.patchVersion = [versionComponents[2] integerValue];
    }
    return version;
}

+ (Class<GADAdNetworkExtras>)networkExtrasClass {
    return [SAAdMobVideoExtra class];
}

- (void)loadRewardedAdForAdConfiguration: (GADMediationRewardedAdConfiguration *)adConfiguration
                       completionHandler: (GADMediationRewardedLoadCompletionHandler)completionHandler {
    _rewardedAd = [[SAAdMobRewardedAd alloc] init];
    [_rewardedAd loadRewardedAdForAdConfiguration:adConfiguration
                                completionHandler:completionHandler];
}

- (void)loadBannerForAdConfiguration:(GADMediationBannerAdConfiguration *)adConfiguration completionHandler:(GADMediationBannerLoadCompletionHandler)completionHandler {
    _bannerAd = [[SAAdMobBannerAd alloc] init];
    [_bannerAd loadBannerForAdConfiguration:adConfiguration completionHandler:completionHandler];
}

- (void)loadInterstitialForAdConfiguration:(GADMediationInterstitialAdConfiguration *)adConfiguration completionHandler:(GADMediationInterstitialLoadCompletionHandler)completionHandler {
    _interstitialAd = [[SAAdMobInterstitialAd alloc] init];
    [_interstitialAd loadInterstitialForAdConfiguration:adConfiguration completionHandler:completionHandler];
}

@end

File: SAAdMobAdapter.swift
//
//  SAAdMobAdapter.swift
//  SuperAwesomeAdMob
//
//  Created by Gunhan Sancar on 19/01/2023.
//

import GoogleMobileAds
import SuperAwesome

@objc(SAAdMobAdapter)
class SAAdMobAdapter: NSObject, GADMediationAdapter {
    
    fileprivate var bannerAd: SAAdMobBannerAd? = nil
    
    required override init() {
        super.init()
    }
    
    static func adapterVersion() -> GADVersionNumber {
        let versionNumber = AwesomeAds.info()?.versionNumber ?? ""
        let versionComponents = versionNumber.components(separatedBy: ".")
        var version = GADVersionNumber()
        if versionComponents.count == 3 {
            version.majorVersion = Int(versionComponents[0]) ?? 0
            version.minorVersion = Int(versionComponents[1]) ?? 0
            version.patchVersion = (Int(versionComponents[2]) ?? 0) * 100
        }
        return version
    }
    
    static func adSDKVersion() -> GADVersionNumber {
        let versionNumber = AwesomeAds.info()?.versionNumber ?? ""
        let versionComponents = versionNumber.components(separatedBy: ".")
        
        if versionComponents.count >= 3 {
            let majorVersion = Int(versionComponents[0]) ?? 0
            let minorVersion = Int(versionComponents[1]) ?? 0
            let patchVersion = Int(versionComponents[2]) ?? 0
            
            return GADVersionNumber(majorVersion: majorVersion,
                                    minorVersion: minorVersion,
                                    patchVersion: patchVersion)
        }
        
        return GADVersionNumber()
    }
    
    static func networkExtrasClass() -> GADAdNetworkExtras.Type? {
        return nil
    }
    
    func loadBanner(for adConfiguration: GADMediationBannerAdConfiguration, completionHandler: @escaping GADMediationBannerLoadCompletionHandler) {
        bannerAd = SAAdMobBannerAd()
        bannerAd?.loadBanner(for: adConfiguration, completionHandler: completionHandler)
    }
    
    func loadInterstitial(for adConfiguration: GADMediationInterstitialAdConfiguration, completionHandler: @escaping GADMediationInterstitialLoadCompletionHandler) {
        
    }
    
    func loadRewardedAd(for adConfiguration: GADMediationRewardedAdConfiguration, completionHandler: @escaping GADMediationRewardedLoadCompletionHandler) {
        
    }
}
